src/service/routing/spec.clj

```diff
@@ -0,0 +1,76 @@
+(ns service.routing.spec
+  (:require [clojure.spec.gen.alpha :as gen]
+            [clojure.spec.alpha :as s]
+            [compojure.api.sweet :refer [context GET resource]]
+            [ring.util.http-response :refer [ok]]
+            [spec-tools.spec :as spec]
+            [service.routing.graph.generators :as g]
+            [service.routing.directions :as dir]
+            ;[expound.alpha :as expound]
+            [clojure.string :as str]
+            [clojure.edn :as edn]))
+            ;[clojure.spec.test.alpha :as stest]))
+
+(s/def ::lon (s/and spec/number? #(<= -180 % 180)))
+(s/def ::lat (s/and spec/number? #(<= -90 % 90)))
+(s/def ::code string?)
+(s/def ::name string?)
+(s/def ::waypoints (s/coll-of ::waypoint-object :kind sequential? :min-count 2))
+(s/def ::waypoint-object (s/keys :req-un [::name ::location]))
+(s/def ::routes (s/coll-of ::route-object :kind sequential?))
+(s/def ::route-object (s/keys :req-un [::geometry ::duration ::distance ::weight ::weight-name ::legs]))
+(s/def ::geometry (s/keys :req-un [::type ::coordinates]))
+(s/def ::coordinate-object (s/tuple ::lon ::lat))
+(s/def ::location ::coordinate-object)
+(s/def ::type string?)
+(s/def ::coordinates (s/coll-of ::coordinate-object :kind sequential?))
+(s/def ::duration spec/number?)
+(s/def ::distance spec/number?)
+(s/def ::weight spec/number?)
+(s/def ::weight-name string?)
+(s/def ::legs vector?)
+(s/def ::waypoint-object (s/keys :req-un [::name ::location]))
+(s/def ::waypoints (s/coll-of ::waypoint-object :kind sequential? :min-count 2))
+(s/def ::geometry (s/keys :req-un [::type ::coordinates]))
+(s/def ::route-object (s/keys :req-un [::geometry ::duration ::distance ::weight ::weight-name ::legs]))
+(s/def ::routes (s/coll-of ::route-object :kind sequential?))
+(s/def ::direction (s/keys :req-un [::code] :opt-un [::waypoints ::routes]))
+
+(def coordinate-regex #"(-?\d+(\.\d+)?),(-?\d+(\.\d+)?)(;(-?\d+(\.\d+)?),(-?\d+(\.\d+)?))+")
+(s/def ::coordinate-regex (s/and string? #(re-matches coordinate-regex %)))
+
+(defn ->coordinates
+  [text]
+  (let [pairs (str/split text #";")]
+    (for [coords pairs]
+      (map edn/read-string (str/split coords #",")))))
+
+;(->coordinates "-122.42,37.78;-77.03,38.91")
+
+(def routes
+  (context "/spec" []
+    :tags ["spec"]
+    :coercion :spec
+
+    (GET "/direction/:coordinates" []
+      :summary "direction with clojure.spec"
+      :path-params [coordinates :- ::coordinate-regex]
+      :return ::direction
+      (ok (let [coords (map zipmap (repeat [:lon :lat])
+                                   (->coordinates coordinates))]
+            (dir/direction (gen/generate (g/graph 1000))
```

Comment:
@mehdisadeghi this is generally not such a good idea. You are creating a graph everytime that a request is made. Of course for testing it was better this way, but now that we have a running server it would be better to have a single graph to query against.

I think for that you can use the [init](https://github.com/weavejester/lein-ring) handler of `lein-ring` and use a `delay` or `future` for its computation such that you dont waste time when starting the server